src/Microsoft.ApplicationInsights/Extensibility/Implementation/Endpoints/EndpointProvider.cs

```diff
@@ -0,0 +1,207 @@
+        /// </summary>
+        /// <returns>Returns the instrumentation key from the connection string.</returns>
+        public string GetInstrumentationKey()
+        {
+            if (this.connectionStringParsed.TryGetValue("InstrumentationKey", out string value))
+            {
+                return value;
+            }
+            else
+            {
+                throw new ConnectionStringMissingInstrumentationKeyException();
+            }
+        }
+
+        /// <summary>
+        /// Parse a connection string and return a Dictionary.
+        /// </summary>
+        /// <remarks>Example: "key1=value1;key2=value2;key3=value3".</remarks>
+        /// <returns>A dictionary parsed from the input connection string.</returns>
+        internal static Dictionary<string, string> ParseConnectionString(string connectionString)
+        {
+            if (connectionString == null)
+            {
+                // TODO: LOG TO ETW ERROR: connection string null
+                throw new ArgumentNullException(nameof(connectionString));
+            }
+
+            var keyValuePairs = connectionString.Split(SplitSemicolon, StringSplitOptions.RemoveEmptyEntries);
+
+            if (keyValuePairs.Length == 0)
+            {
+                // TODO: LOG TO ETW ERROR: connection string empty
+                throw new ArgumentException("Connection string cannot be empty.");
+            }
+
+            var dictionary = new Dictionary<string, string>(keyValuePairs.Length, StringComparer.OrdinalIgnoreCase);
+
+            foreach(var pair in keyValuePairs)
+            {
+                var keyAndValue = pair.Split('=');
+                if (keyAndValue.Length != 2)
+                {
+                    // TODO: LOG TO ETW ERROR: connection string invalid format
+                    throw new ConnectionStringInvalidDelimiterException("The Connection String has invalid formatting and cannot be parsed. Expected: 'key1=value1;key2=value2;key3=value3'");
+                }
+
+                if (dictionary.ContainsKey(keyAndValue[0]))
+                {
+                    // TODO: LOG TO ETW ERROR: connection string duplicate key
+                    throw new ConnectionStringDuplicateKeyException($"The Connection String has a duplicate key '{keyAndValue[0]}'.");
+                }
+
+                dictionary.Add(keyAndValue[0], keyAndValue[1]);
+            }
+
+            return dictionary;
+        }
+
+        /// <summary>
+        /// Construct a Uri from the possible parts.
+        /// Will also attempt to sanitize user input.
+        /// </summary>
+        /// <remarks>
+        /// Format: "location.prefix.suffix".
+        /// Example: "https://westus2.dc.applicationinsights.azure.cn/".
+        /// </remarks>
+        /// <returns>Returns a <see cref="Uri"/> built from the inputs.</returns>
+        internal static Uri BuildUri(string prefix, string suffix, string location = null)
```

Comment:
bool TryBuildUri(..., out var uri) and rather return false. And above, where you call this method check the return value